### PR TITLE
WD-693 - Refactor the project

### PR DESCRIPTION
**Pull request by goulinkh**
## Description

The Current behavior of the `jujulib` for facade classes is to return back the class of the request facade if the **exact** version is supported by the API, otherwise return `undefined`.
This is causing issues now, where the latest version of facades aren't working on some Juju controllers which result in having the dashboard failing everywhere, in order to fix this issue we would like to fallback to the **latest** supported version by the controller.

### Advantages of this approach

The dashboard will work in all the cases.

### Limitations

If the latest version of the juju dashboard is using a specific functionality that is only available on the latest version of a facade, `jujulib` may return an older version of the facade which doesn't support the functionality that we use.

Which is better that not having the dashboard to work at all!

## Done

There will be part 2 PR that will specifically address the issue mentioned above. This PR is a refactor of the project to make the second PR feasible.

- [x] Better support for Typescript: Mainly Typescript file in the root of `/api`
- [x] Change the structure of the facades: e.g. `/api/facades/ClientV5` -> `/api/facades/client/ClientV5.ts`
- [x] Change the structure of the generator codebase: generator related code was in separate folders in the root of the project, I find the new structure simpler to user and read
- [x] Remove the docs from the project: Each time we generate the facades the docs files get updated, which cause a lot of Git changes to track, instead we can make the docs generation a part of the Github action that deploys the `github.io` pages
- [x] Add a script `reganarate-all.js`: that generates all the facades based on historical `schema.json` from Github
- [x] Remove the date of generating the facades from the header

## TODO

- [ ] File a PR to build the docs with the deployment the Github page
- [ ] File a PR to get the most recent version of the facades that is supported by the Juju controller

## QA

- `npm run build`
- `yarn link`
- Update the imports in the juju-dashboard to work with the new structure
- Make sure that the dashboard is working as expected

### CODE REVIEW

**Comment by goulinkh** [comment]
I'm sorry for the big changes in the PR @huwshimi, I felt that the changes where all related! there are 2 commits:
- the first one removes the old files + structure
- the second one adds the new structures + TS improvement

File: generator/schema/regenerate-all.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    await downloadFile(schema.schema, join(baseFolder, "schema.json"))
+    await fs.promises.writeFile(
+      join(baseFolder, "juju-version.txt"),
+      schema["juju-version"]
+    )
+    await fs.promises.writeFile(
+      join(baseFolder, "juju-git-sha.txt"),
+      schema["juju-git-sha"]
+    )
+    await execShellCommand("npm run generate-facades")
```
Comment (huwshimi) [inline generator/schema/regenerate-all.js:68]: I think we probably want to start using `yarn` for this project as well.

**Review by huwshimi — COMMENTED** [review]
I'm having trouble getting this to work with Juju Dashboard.

I cleaned both projects and linked jujulib in the dashboard but I get this error when I run `yarn start` from the dashboard:

```
Failed to compile.

Module not found: Error: Can't resolve './facades/admin/AdminV3' in '/home/ubuntu/code/js-libjuju/dist/api'
Did you mean 'AdminV3.js'?
BREAKING CHANGE: The request './facades/admin/AdminV3' failed to resolve only because it was resolved as fully specified
(probably because the origin is strict EcmaScript Module, e. g. a module with javascript mimetype, a '*.mjs' file, or a '*.js' file where the package.json contains '"type": "module"').
The extension in the request is mandatory for it to be fully specified.
Add the extension to the request.
ERROR in ../js-libjuju/dist/api/client.js 13:0-46
Module not found: Error: Can't resolve './facades/admin/AdminV3' in '/home/ubuntu/code/js-libjuju/dist/api'
Did you mean 'AdminV3.js'?
BREAKING CHANGE: The request './facades/admin/AdminV3' failed to resolve only because it was resolved as fully specified
(probably because the origin is strict EcmaScript Module, e. g. a module with javascript mimetype, a '*.mjs' file, or a '*.js' file where the package.json contains '"type": "module"').
The extension in the request is mandatory for it to be fully specified.
Add the extension to the request.

webpack compiled with 1 error
```
It looks like this could be related:
https://github.com/facebook/create-react-app/issues/11865#issuecomment-1059714419

It's possible we need to use WebPack to build this project into commonjs modules or something.

File: generator/schema/regenerate-all.js (hunk 1 above)
Comment (goulinkh) [reply generator/schema/regenerate-all.js]: yes, also I should add this as a yarn script!

File: .eslintrc [hunk 2]
```diff
@@ -2,15 +2,13 @@ env:
...
+  ecmaVersion: 2018
 rules:
   eqeqeq: ["error"]
   no-mixed-spaces-and-tabs: ["error"]
   no-trailing-spaces: ["error"]
   no-undef: 2
   no-unused-vars: [2, {args: "none"}]
   no-unreachable: 2
-  quotes: [2, "single"]
-  semi: [2, "always"]
```
[the remark this answers is not in the document]
Comment (goulinkh) [reply .eslintrc:16]: Makes sense, apparently this file wasn't getting picked up by eslint, so I updated the eslint rules with TS support (although I disabled a lot of rules).

**Comment by goulinkh** [comment]
@huwshimi I fixed the import issue by adding `.js` in some imports in the lib, should be good for now!
BTW here are my imports:
```ts
import actions from "@canonical/jujulib/dist/api/facades/action/ActionV7";
import allWatcher from "@canonical/jujulib/dist/api/facades/all-watcher/AllWatcherV3";
import annotations from "@canonical/jujulib/dist/api/facades/annotations/AnnotationsV2";
import applications from "@canonical/jujulib/dist/api/facades/application/ApplicationV15";
import client from "@canonical/jujulib/dist/api/facades/client/ClientV6";
import cloud from "@canonical/jujulib/dist/api/facades/cloud/CloudV7";
import controller from "@canonical/jujulib/dist/api/facades/controller/ControllerV11";
import modelManager from "@canonical/jujulib/dist/api/facades/model-manager/ModelManagerV9";
import pinger from "@canonical/jujulib/dist/api/facades/pinger/PingerV1";
```

File: README.md [hunk 3]
```diff
@@ -9,28 +8,28 @@
...
 ```javascript
-import ApplicationV5 from "jujulib/api/facades/application-v5.js";
-import ClientV1 from "jujulib/api/facades/client-v1.js";
-import ClientV2 from "jujulib/api/facades/client-v2.js";
-
-const juju = await jujulib.connect(
-  'wss://example.com/model/x-y-z/api',
-  {facades: [ApplicationV5, ClientV1, ClientV2]}
-);
+import ApplicationV5 from "jujulib/api/facades/application-v5.js"
```
Comment (huwshimi) [inline README.md:26]: It looks like these import examples are out of date now.

File: README.md [hunk 4]
```diff
@@ -186,9 +185,9 @@ The Juju facade API files are generated from a supplied Juju schema.
 
 To generate this schema you will need to clone the [Juju repository](https://github.com/juju/juju/) and then run `go run github.com/juju/juju/generate/schemagen -admin-facades --facade-group=client,jimm ./apiserver/facades/schema.json` to generate a schema file that contains the publicly available facades as well as the set of facades for JAAS. Other `--facade-group` options are `latest` and `all`.
 
-After generating a new schema run `npm run store-schema` which will store the updated schema and necessary meta data in this project.
+After generating a new schema run `yarn run store-schema` which will store the updated schema and necessary meta data in this project.
```
Comment (huwshimi) [inline README.md:188]: Thanks for update these!

File: api/client.ts [hunk 5]
```diff
@@ -7,25 +7,30 @@
...
-  controllerTag: string;
-  serverVersion: string;
-  servers: object[];
-  user: object;
-  getFacade: (string) => Facade;
+  controllerTag?: string;
+  serverVersion?: string;
+  servers?: object[];
+  user?: object;
+  getFacade?: (name: string) => Facade;
```
Comment (huwshimi) [inline api/client.ts:33]: Thanks for improving all these types.

File: api/client.ts [hunk 6]
```diff
@@ -100,15 +107,13 @@ function connect(
...
+    const ws = new options!.wsclass!(url);
     const handler = createAsyncHandler(callback, resolve, reject);
-    ws.onopen = (evt) => {
-      handler(null, new Client(ws, options));
+    ws.onopen = (_evt) => {
+      handler(null, new Client(ws, options!));
     };
     ws.onclose = (evt) => {
-      handler("cannot connect WebSocket: " + evt.reason, null);
+      handler("cannot connect WebSocket: " + evt.reason, undefined);
```
Comment (huwshimi) [inline api/client.ts:116]: It looks like `value` is optional here so you could probable just remove `undefined` from this callback.

File: generator/templates/facade.js [hunk 7]
```diff
@@ -0,0 +1,107 @@
+export default function generateFacadeTemplate(facadeTemplate) {
```
Comment (huwshimi) [inline generator/templates/facade.js:1]: I suspect this file wasn't meant to be committed.

File: generator/templates/readme.js [hunk 8]
```diff
@@ -0,0 +1,97 @@
+export default (r) => {
```
Comment (huwshimi) [inline generator/templates/readme.js:1]: I suspect this file shouldn't have been committed.

File: api/helpers.ts [hunk 9]
```diff
@@ -6,17 +6,17 @@
...
+import { createAsyncHandler } from "./utils.js";
 
 /**
   Decorate the Admin facade class.
 
   @param {Object} cls The auto-generated class.
   @returns {Object} The decorated class.
 */
-function wrapAdmin(cls) {
+function wrapAdmin(cls: any): object {
```
Comment (huwshimi) [inline api/helpers.ts:19]: Over time we can improve these `any`, `object`, `Function` etc. types.

**Review by huwshimi — APPROVED** [review]
This is a great step in the right direction!

File: api/helpers.ts (hunk 9 above)
Comment (goulinkh) [reply api/helpers.ts:19]: yes! I've a created an issue to track this: https://github.com/juju/js-libjuju/issues/93

**Pull request merged**